File: algorithms/easy/remove-linked-list-elements.py
# Definition for singly-linked list.
class ListNode(object):
    def __init__(self, x, nt=None):
        self.val = x
        self.next = nt


class Solution(object):
    def removeElements(self, head, val):
        """
        :type head: ListNode
        :type val: int
        :rtype: ListNode
        """
        pre = ListNode(None)
        pre.next = head
        cur = head
        head = pre
        while pre and pre.next:
            if cur.val == val:
                pre.next = cur.next
                cur = cur.next
            else:
                pre = pre.next
                cur = pre and pre.next
        return head.next


def test():
    head = ListNode(1, ListNode(2, ListNode(3, ListNode(4, ListNode(3)))))
    Solution().removeElements(head, 3)
    l = []
    while head:
        l.append(head.val)
        head = head.next
    print l
    assert l == [1, 2, 4]

if __name__ == '__main__':
    test()

File: algorithms/easy/remove-linked-list-elements.c
#include <stdio.h>
#include <stdlib.h>

struct ListNode {
    int val;
    struct ListNode *next;
};

struct ListNode* newNode(int val, struct ListNode *next) {
    struct ListNode *node = malloc(sizeof(struct ListNode));
    node->val = val;
    node->next = next;
    return node;
}

struct ListNode* removeElements(struct ListNode* head, int val) {
    while (head != NULL && head->val == val) {
        head = head->next;
    }
    if (head == NULL)
        return NULL;
    struct ListNode *cur = head;
    while (cur->next != NULL) {
        if (cur->next->val == val) {
            cur->next = cur->next->next;
        } else {
            cur = cur->next;
        }
    }
    return head;
}

int main() {
    struct ListNode *l1 = newNode(1, newNode(2, newNode(2, newNode(4, newNode(5, NULL)))));
    struct ListNode *l2 = removeElements(l1, 2);
    while(l2 != NULL) {
        printf("%d ", l2->val);
        l2 = l2->next;
    }
    printf("\n");

    l1 = newNode(1, newNode(1, newNode(1, newNode(4, newNode(5, NULL)))));
    l2 = removeElements(l1, 1);
    while(l2 != NULL) {
        printf("%d ", l2->val);
        l2 = l2->next;
    }
    printf("\n");

    l1 = newNode(1, newNode(1, newNode(1, newNode(1, newNode(1, NULL)))));
    l2 = removeElements(l1, 1);
    while(l2 != NULL) {
        printf("%d ", l2->val);
        l2 = l2->next;
    }
    printf("\n");

    l1 = newNode(1, newNode(1, newNode(1, newNode(1, newNode(1, NULL)))));
    l2 = removeElements(NULL, 1);
    while(l2 != NULL) {
        printf("%d ", l2->val);
        l2 = l2->next;
    }
    printf("\n");
}
